### ISSUE TITLE
PdAudioController on OSX

### ISSUE THREAD

**Issue by designerfuzzi** [libpd/libpd#93]
i try to implement libpd inside an OSX bundle, when it comes to #import "PdAudioController.h" it uses Foundation Framework and AVFoundation of iOS and the whole Controller is written for iOS.
What to change it can be used in a OSX bundle instead of iOS?

**Comment by danomatika** [comment]
Pinging @richardeakin

PdAudioController & PdAudioUnit don't currently work on OS X. There is a `cocoa` branch with some work into it, but one main issue is getting a working aggregate audio unit with both and input *and* output is not as simple as on iOS. Anyway, there are some samples in `samples/objc/osx` but they currently do not work. I haven't checked it in a while and they may not build either :P

**Comment by danomatika** [comment]
TO follow up, you can leave out PdAudioUnit & PdAudiCOntroller and libpd will build fine. In that case, of course you'll need to manually handle audio samples and calling PdBase processFloat, but it will work. The audio controller & unit are convenience classes but not required.

**Comment by richardeakin** [comment]
PdAudioController was designed for iOS, though PdAudioUnit is more straightforward to update for Cocoa (it will also need a couple changes around the HAL output unit).  I'm sorry, I don't have time to work on this these days, though the cocoa branch is a good place to start.. you're best bet would be to use your own audio unit, and either libpd's Obj-C or C++ wrapper in that.

**Comment by designerfuzzi** [comment]
So PdAudioController is really just for handling iOS specific Audio in Case of interuption by phonecalls and so on. So making my own Audio handling is not that complex, just need to know the structure of dataflow with PD based synthesis, in example Audio Input Entrypoint and Audio Output Buffer, right? Where to look into exactly? As you suggest, starting with seeking for what i search in PdAudioUnit?
In my case handling AudioSynthesis is not that important, but getting most of the values used in processing it is nice to bind it to visual art work. And maybe getting some infos out of the resulting audiostream with analysing via FFT or so.. 
Main target is to enable SyncJams based on libpd 'sessions' being visualized in OSX Quartz Composer Environment, as discussed last weekend at midihack.com berlin.

**Comment by danomatika** [comment]
Not really. The PdAudioController handles setting up and running the PdAudioUnit. The PdAudioUnit handles audio input & output and calls the libpd process functions.

If you don't need audio and just want to use messaging with pd/libpd, you can just call one of the PdBase process methods with a dummy buffer that has no data. This is similar to what the basic c pdtest does.

**Comment by shakfu** [comment]
I wish I read this issue before I started trying to use PdAudioController on Mac OS X

**Comment by danomatika** [comment]
We haven't provided a macOS version of PdAudioController/PdAudioUnit as existing frameworks like Portaudio and RtMidi were already available. When this project was started, audio programming on iOS was relatively new, so there needed to be a built-in way to access audio IO.

Using Portaudio or RtMidi is relatively easy. Once you have the audio streams set up, process samples into/out of libpd with the PdBase process* methods or the libpd process* functions. Unlike iOS, Core Audio on macOS seems pretty good at providing the right buffer sizes with a multiple of Pd's 64 frame block size, so you can just process samples directly inside the audio stream IO function:

~~~
// Obj-C
int ticks = bufferFrames / [PdBase getBlockSize];
[PdBase processFloatWithInputBuffer:inputBuffer outputBuffer:outputBuffer ticks:ticks];

// C
int ticks = bufferFrames / libpd_blocksize();
libpd_process_float(ticks, inputBuffer, outputBuffer);
~~~

Eventually I might refactor out a Portaudio + libpd Obj-C wrapper I made for a project into an example here.

**Comment by danomatika** [comment]
Also, there is a C++ RtMidi example here: https://github.com/libpd/libpd/tree/master/samples/cpp/pdtest_rtaudio

**Comment by shakfu** [comment]
Thanks very much. I got the python/pyaudio example after making it python3. Hearing a sound at the end of the process made a huge impact!

I then tried to do something similar and learn swift and xcode by wrapping the objc example but then started hitting a wall with PdAudioController. Looks like I need to get my head around portaudio and core audio...

> Eventually I might refactor out a Portaudio + libpd Obj-C wrapper I made for a project into an example here.

That would be amazing. Examples with sound are great! Thanks for your help and advice!
